### PR TITLE
Add conformance tests for new-expr [stream]

### PR DESCRIPTION
**Pull request by KRVPerera**
## Purpose
Add conformance tests for new-expr stream covering [StreamImplementor](https://ballerina.io/spec/lang/master/#StreamImplementor)

### CODE REVIEW

**Comment by KRVPerera** [comment]
Need to resolve https://github.com/ballerina-platform/ballerina-spec/issues/1036 to add more test cases.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+Description: Test new-expression to create a stream without type descriptors.
+Labels: new-expr, stream-type-descriptor
+
+function init() {
+    var _ = new stream<>(); // @error missing type descriptor
+}
+
+Test-Case: error
+Description: Test new-expression to create a stream with incompatible type-descriptor.
+Labels: new-expr, stream-type-descriptor
```
Comment (lochana-chathura) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:11]: missing string label

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+function init() {
+    stream<int> _ = new stream<string>(); // @error incompatible types: expected 'stream<int>', found 'stream<string>'
+}
+
+Test-Case: output
+Description: Test new-expression to create a stream with incompatible type-descriptor.
+Fail-Issue: ballerina-platform/ballerina-lang/issues/34963
```
Comment (lochana-chathura) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:19]:
```suggestion
Fail-Issue: ballerina-platform/ballerina-lang#34963
```

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+Labels: new-expr, stream-type-descriptor
```
Comment (lochana-chathura) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:20]: missing labels `error-type`, `optional-type`

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+function init() {
+    stream s = new stream<string, error?>();
+    io:println(s); // @output stream <stream<int,error?>>
+}
+
+Test-Case: output
+Description: Test new-expression to create a empty streams.
+Labels: new-expr, stream-type-descriptor, int-type-descriptor
```
Comment (lochana-chathura) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:29]: int is in the minimal subset. missing labels: record-type, is-expr, if-else-stmt, module-type-defn, error-type, optional-type, union-type. check other places also.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+function init() {
+    stream<int, error?> emptyStream = new stream<int, error?>();
+    int|error? d = getRecordValue(emptyStream.next());
+    io:println(d); // @output
+}
+
+Test-Case: output
+Description: Test new-expr to create a stream without completion type
```
Comment (lochana-chathura) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:50]:
```suggestion
Description: Test new-expr to create a stream without completion type.
```

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+Test-Case: output
+Description: Test closing of stream created with new-expression.
+Labels: new-expr, stream-type-descriptor, int-type-descriptor
+
+type IntRecord record {|
+    int value;
+|};
+
+function getRecordValue(IntRecord|error? returnedVal) returns int|error? {
```
Comment (gimantha) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:112]: We need few tests for the StreamImplementor class with close methods too.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    int|error? a = getRecordValue(numberStream.next());
+    int|error? b = getRecordValue(numberStream.next());
+    io:println(a); // @output 1
+    io:println(b); // @output 2
+    error? closed = numberStream.close();
+    io:println(closed is ()); // @output true
+}
+
+Test-Case: panic
+Description: Test closed stream next.
```
Comment (gimantha) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:140]: Shall we check the auto-closing behavior as well?

Comment (gimantha) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:140]: for example, calling, `next` after `next` returns (), should panic

Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:140]: I am waiting on https://github.com/ballerina-platform/ballerina-spec/issues/1036

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt (hunk 4 above)
Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:29]: Thanks I think I have not focused on labels

**Review by MaryamZi — COMMENTED** [review]
I think we can add quite a few more tests.

e.g.,
when the class-reference refers to a stream type
- having more than one argument
- the single argument not being a stream implementor
- incompatibilities between the stream-type and the stream impementor next method's return type
- the line "The explicit-new-expr is isolated if the the arg-list is empty or if the expression for the single argument is isolated."

And similar tests where applicable for implicit-new-expr.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+function init() {
+    stream<int> emptyStream = new stream<int>();
+    int? d = getRecordValue(emptyStream.next());
+    io:println(d); // @output
+}
+
+Test-Case: output
+Description: Test new-expression with int streams.
```
[the remark this answers is not in the document]
Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:72]: Will update

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt (hunk 7 above)
Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:140]: Spec issue is updated with details

Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:139]: Tracked by https://github.com/ballerina-platform/ballerina-lang/issues/34934

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+Labels: error, optional-type, new-expr, string, stream-type-descriptor
+
+function init() {
+    stream s = new stream<string, error?>();
+    io:println(s); // @output stream <stream<string,error?>>
+}
+
+Test-Case: output
+Description: Test new-expression to create a empty streams.
+Labels: int, record-type, optional-type, is-expr, module-type-defn, new-expr, union-type, stream-type-descriptor
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:28]: `null` label is missing

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+
+function init() {
+    stream<int, error?> emptyStream = new stream<int, error?>();
+    IntRecord|error? d = emptyStream.next();
+    io:println(d is ()); // @output true
+}
+
+Test-Case: output
+Description: Test new-expr to create a stream without completion type.
+Labels: int, record-type, optional-type, new-expr, is-expr, stream-type-descriptor
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:42]: `nil-type` label is missing

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+
+function init() {
+    stream<int> emptyStream = new stream<int>();
+    IntRecord? d = emptyStream.next();
+    io:println(d is ()); // @output true
+}
+
+Test-Case: error
+Description: Test new-expr to create a stream with stream extra parameters.
+Labels: int, error, optional-type, new-expr, module-class-defn, stream-type-descriptor
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:56]: `string` label is missing.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+
+function init() {
+    NumbersGenerator numbersGenerator = new;
+    stream<string, error?> numberStream = new stream<string, error?>(numbersGenerator, string); // @error 'new(numbersGenerator, null)' is not a valid constructor for streams type
+}
+
+
+Test-Case: error
+Description: Test new-expr to create a stream with stream implementor with next method with incompatible return type.
+Labels: int, error, optional-type, new-expr, module-class-defn, stream-type-descriptor
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:74]: `return-stmt`, `isolated-function` labels missing.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+function init() {
+    NumbersGenerator numbersGenerator = new;
+    stream<int, error?> numberStream = new stream<int, error?>(numbersGenerator); // @error invalid stream constructor. expected a subtype of 'object { public isolated function next() returns (record {| int value; |}|error)?; }', but found 'NumbersGenerator'
+}
+
+Test-Case: output
+Description: Test new-expression with int stream created via a stream with parameters and implementor.
+Labels: int, record-type, optional-type, explicit-new-expr, module-type-defn, module-class-defn, if-else-stmt, is-expr, union-type, error, stream-type-descriptor
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:105]: `return-stmt`, `field-access-expr`

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+    int|error? a = getRecordValue(numberStream.next());
+    io:println(a); // @output 1
+    int|error? b = getRecordValue(numberStream.next());
+    io:println(b); // @output 2
+}
+
+Test-Case: output
+Description: Test close method return type of stream created with stream implementor not implementing close.
+Labels: int, error, record-type, optional-type, if-else-stmt, is-expr, module-type-defn, module-class-defn, union-type, explicit-new-expr, stream-type-descriptor
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:138]: `return-stmt`, `field-access-expr` are missing.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+    stream<int, error?> newStream = new stream<int, error?>();
+    error? closed = newStream.close();
+    var _ = newStream.next(); // @panic
+}
+
+Test-Case: panic
+Description: Test auto closing of stream after completion type.
+// Fail-Issue: ballerina-platform/ballerina-lang/issues/34934
+Labels: int, record-type, optional-type, if-else-stmt, is-expr, module-type-defn, module-class-defn, new-expr, stream-type-descriptor
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:185]: `return-stmt`, `field-access-expr`, `nil-type`, 'function-defn`, `function-call-expr` labels are missing.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 16]
```diff
@@ -0,0 +1,303 @@
...
+    io:println(value is ()); // @output true
+    IntRecord? panicVar = newStream.next(); // @panic
+    io:println(panicVar is IntRecord); // @output false
+}
+
+Test-Case: panic
+Description: Test closed stream next when stream implementor does not have close method.
+// Fail-Issue: ballerina-platform/ballerina-lang/issues/34934
+Labels: int, record-type, error, optional-type, union-type, is-expr, module-type-defn, module-class-defn, explicit-new-expr, stream-type-descriptor
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:225]: `return-stmt`, `nil-type` labels are missing

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+    error? closed = newStream.close();
+    io:println(closed is ()); // @output true
+    IntRecord|error? panicVar = newStream.next(); // @panic
+    io:println(panicVar is IntRecord); // @output false
+}
+
+Test-Case: output
+Description: Test closed stream next with stream implementor with close method.
+Labels: int, record-type, error, optional-type, union-type, module-type-defn, module-class-defn, if-else-stmt, is-expr, explicit-new-expr, stream-type-descriptor
+
```
Comment (SasinduDilshara) [inline conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:250]: `return-stmt`, `nil-type`, `field-access-expr` labels are missing.

Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:263]: Thanks. As of now `return-stmt`, `function-defn`, `function-call-expr` all are in the minimal subset and I am not testing them specifically. 
I added `nil-type`, `method-call-expr`, `member-access-expr`

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt (hunk 15 above)
Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:194]: Thanks

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt (hunk 12 above)
Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt]: Thanks no need to add `return-stmt` as in minimal subset and not testing it specifically here.

File: conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt (hunk 13 above)
Comment (KRVPerera) [reply conformance/lang/expressions/new-expr/explicit-new-expr/stream_class_descriptor.balt:109]: Thanks

**Comment by KRVPerera** [comment]
> I think we can add quite a few more tests.
> 
> e.g., when the class-reference refers to a stream type
> 
> * having more than one argument
> * the single argument not being a stream implementor
> * incompatibilities between the stream-type and the stream impementor next method's return type
> * the line "The explicit-new-expr is isolated if the the arg-list is empty or if the expression for the single argument is isolated."
> 
> And similar tests where applicable for implicit-new-expr.

Need to address last point others are fixed

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 18]
```diff
@@ -0,0 +1,317 @@
+Test-Case: error
+Description: Test new-expression to create a stream without stream type parameters.
+Labels: explicit-new-expr, stream, var
+
+function errorFunction() {
+    var _ = new stream<>(); // @error missing type parameters
+}
+
```
Comment (pcnfernando) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:8]: Can we include tests for
```
    stream<int> streamName = new;
```

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+type IntRecord record {|
+    int value;
+|};
+
+function init() {
+    stream<int, error?> emptyStream = new stream<int, error?>();
+    IntRecord|error? d = emptyStream.next();
+    io:println(d is ()); // @output true
+}
+
```
Comment (pcnfernando) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:40]: Few tests with parser errors

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+Test-Case: error
+Description: Test new-expression to create a stream with incompatible stream type parameters.
+Labels: explicit-new-expr, stream-type-parameters, string
+
+function errorFunction() {
+    stream<int> _ = new stream<string>(); // @error incompatible types: expected 'stream<int>', found 'stream<string>'
+}
+
+Test-Case: output
+Description: Test new-expression to create a stream with incompatible type-descriptor.
```
Comment (MaryamZi) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:18]: Is this description correct?

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+    int i = 0;
+    public isolated function next() returns string|error? {
+        self.i += 1;
+        return self.i.toBalString();
+    }
+}
+
+function errorFunction() {
+    NumbersGenerator numbersGenerator = new;
+    stream<string, error?> numberStream = new stream<string, error?>(numbersGenerator, string); // @error 'new(numbersGenerator, null)' is not a valid constructor for streams type
```
Comment (MaryamZi) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:70]: The error is not correct? Has null in the error.

1. Let's create an issue for jBallerina
2. Let's fix the error here to reflect the ideal error message

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+    int i = 0;
+    public isolated function next() returns string|error? {
+        self.i += 1;
+        return self.i.toBalString();
+    }
+}
+
+function errorFunction() {
+    NumbersGenerator numbersGenerator = new;
+    stream<int, error?> numberStream = new stream<int, error?>(numbersGenerator); // @error incompatible types: expected '(IntRecord|ErrorB)?', found '(record {| int value; |}|ErrorA)?'
```
Comment (MaryamZi) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:88]: Here again the error message is not clear, right? We should use the proper error to assert and also create issues for jBallerina to improve the errors.

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+type IntRecord record {|
+    int value;
+|};
+
+function getRecordValue(IntRecord|error? returnedVal) returns int|error? {
+    if returnedVal is IntRecord {
+        return returnedVal.value;
+    } else {
+        return returnedVal;
+    }
```
Comment (MaryamZi) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:120]:
```suggestion
    }
    return returnedVal;
```

Comment (MaryamZi) [reply conformance/lang/expressions/new-expr/stream_class_descriptor.balt:120]: If we are changing this, applicable in other places too.

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+    io:println(a); // @output 1
+    int|error? b = getRecordValue(numberStream.next());
+    io:println(b); // @output 2
+    error? closed = numberStream.close();
+    io:println(closed is ()); // @output true
+}
+
+Test-Case: panic
+Description: Test closed stream next.
+// Fail-Issue: ballerina-platform/ballerina-lang/issues/34934
```
Comment (MaryamZi) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:184]: Why `//`?

File: conformance/lang/expressions/new-expr/stream_class_descriptor.balt [hunk 25]
```diff
@@ -0,0 +1,317 @@
...
+    io:println(b); // @output 2
+    stream<int, error?> numberStreamI = new (numbersGenerator);
+    int|error? ai = getRecordValue(numberStreamI.next());
+    io:println(ai); // @output 3
+    int|error? bi = getRecordValue(numberStreamI.next());
+    io:println(bi); // @output 4
+}
+
+Test-Case: output
+Description: Test close method return type of stream created with stream implementor not implementing close.
```
Comment (MaryamZi) [inline conformance/lang/expressions/new-expr/stream_class_descriptor.balt:146]: Shall we add an error test with incompatible types?

**Review by MaryamZi — COMMENTED** [review]
We need to add test with `var` also.

**Comment by gimantha** [comment]
> > > I think we can add quite a few more tests.
> > > e.g., when the class-reference refers to a stream type
> > > 
> > > * having more than one argument
> > > * the single argument not being a stream implementor
> > > * incompatibilities between the stream-type and the stream impementor next method's return type
> > > * the line "The explicit-new-expr is isolated if the the arg-list is empty or if the expression for the single argument is isolated."
> > > 
> > > And similar tests where applicable for implicit-new-expr.
> > 
> > 
> > Need to address last point others are fixed
> 
> This still needs to be addressed?

Last point needs to be addressed. I will push the changes

**Comment by gimantha** [comment]
Closing till we address the review suggestion and add tests for isolated-ness

**Pull request closed without merging**